File: 2024/day-04/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	// Read input file
	infile := "input.txt"
	if len(os.Args) >= 2 {
		infile = os.Args[1]
	}

	file, err := os.Open(infile)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	defer file.Close()

	// Parse the grid
	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	R := len(grid)
	C := len(grid[0])
	p1, p2 := 0, 0

	// Process the grid for patterns
	for r := 0; r < R; r++ {
		for c := 0; c < C; c++ {
			// Check for "XMAS" horizontally, vertically, and diagonally
			if c+3 < C && grid[r][c] == 'X' && grid[r][c+1] == 'M' && grid[r][c+2] == 'A' && grid[r][c+3] == 'S' {
				p1++
			}
			if r+3 < R && grid[r][c] == 'X' && grid[r+1][c] == 'M' && grid[r+2][c] == 'A' && grid[r+3][c] == 'S' {
				p1++
			}
			if r+3 < R && c+3 < C && grid[r][c] == 'X' && grid[r+1][c+1] == 'M' && grid[r+2][c+2] == 'A' && grid[r+3][c+3] == 'S' {
				p1++
			}

			// Check for "SAMX" horizontally, vertically, and diagonally
			if c+3 < C && grid[r][c] == 'S' && grid[r][c+1] == 'A' && grid[r][c+2] == 'M' && grid[r][c+3] == 'X' {
				p1++
			}
			if r+3 < R && grid[r][c] == 'S' && grid[r+1][c] == 'A' && grid[r+2][c] == 'M' && grid[r+3][c] == 'X' {
				p1++
			}
			if r+3 < R && c+3 < C && grid[r][c] == 'S' && grid[r+1][c+1] == 'A' && grid[r+2][c+2] == 'M' && grid[r+3][c+3] == 'X' {
				p1++
			}
			if r-3 >= 0 && c+3 < C && grid[r][c] == 'S' && grid[r-1][c+1] == 'A' && grid[r-2][c+2] == 'M' && grid[r-3][c+3] == 'X' {
				p1++
			}
			if r-3 >= 0 && c+3 < C && grid[r][c] == 'X' && grid[r-1][c+1] == 'M' && grid[r-2][c+2] == 'A' && grid[r-3][c+3] == 'S' {
				p1++
			}

			// Check for "MAS" patterns surrounded by "M/S"
			if r+2 < R && c+2 < C && grid[r][c] == 'M' && grid[r+1][c+1] == 'A' && grid[r+2][c+2] == 'S' && grid[r+2][c] == 'M' && grid[r][c+2] == 'S' {
				p2++
			}
			if r+2 < R && c+2 < C && grid[r][c] == 'M' && grid[r+1][c+1] == 'A' && grid[r+2][c+2] == 'S' && grid[r+2][c] == 'S' && grid[r][c+2] == 'M' {
				p2++
			}
			if r+2 < R && c+2 < C && grid[r][c] == 'S' && grid[r+1][c+1] == 'A' && grid[r+2][c+2] == 'M' && grid[r+2][c] == 'M' && grid[r][c+2] == 'S' {
				p2++
			}
			if r+2 < R && c+2 < C && grid[r][c] == 'S' && grid[r+1][c+1] == 'A' && grid[r+2][c+2] == 'M' && grid[r+2][c] == 'S' && grid[r][c+2] == 'M' {
				p2++
			}
		}
	}

	// Output results
	fmt.Println(p1)
	fmt.Println(p2)
}

File: 2024/day-04/main.rs
use std::env;
use std::fs;

fn main() {
    // Read input file
    let args: Vec<String> = env::args().collect();
    let infile = args.get(1).unwrap_or(&"4.in".to_string());
    let grid: Vec<String> = fs::read_to_string(infile)
        .expect("Could not read file")
        .lines()
        .map(String::from)
        .collect();

    let rows = grid.len();
    let cols = grid[0].len();
    let mut p1 = 0;
    let mut p2 = 0;

    for r in 0..rows {
        for c in 0..cols {
            // Check for "XMAS" horizontally, vertically, and diagonally
            if c + 3 < cols && is_pattern(&grid, r, c, &['X', 'M', 'A', 'S'], (0, 1)) {
                p1 += 1;
            }
            if r + 3 < rows && is_pattern(&grid, r, c, &['X', 'M', 'A', 'S'], (1, 0)) {
                p1 += 1;
            }
            if r + 3 < rows && c + 3 < cols && is_pattern(&grid, r, c, &['X', 'M', 'A', 'S'], (1, 1)) {
                p1 += 1;
            }
            if c + 3 < cols && is_pattern(&grid, r, c, &['S', 'A', 'M', 'X'], (0, 1)) {
                p1 += 1;
            }
            if r + 3 < rows && is_pattern(&grid, r, c, &['S', 'A', 'M', 'X'], (1, 0)) {
                p1 += 1;
            }
            if r + 3 < rows && c + 3 < cols && is_pattern(&grid, r, c, &['S', 'A', 'M', 'X'], (1, 1)) {
                p1 += 1;
            }
            if r >= 3 && c + 3 < cols && is_pattern(&grid, r, c, &['S', 'A', 'M', 'X'], (-1, 1)) {
                p1 += 1;
            }
            if r >= 3 && c + 3 < cols && is_pattern(&grid, r, c, &['X', 'M', 'A', 'S'], (-1, 1)) {
                p1 += 1;
            }

            // Check for "MAS" patterns surrounded by "M/S"
            if r + 2 < rows && c + 2 < cols {
                if is_surrounded_pattern(&grid, r, c, &['M', 'A', 'S'], 'M', 'S') {
                    p2 += 1;
                }
                if is_surrounded_pattern(&grid, r, c, &['M', 'A', 'S'], 'S', 'M') {
                    p2 += 1;
                }
                if is_surrounded_pattern(&grid, r, c, &['S', 'A', 'M'], 'M', 'S') {
                    p2 += 1;
                }
                if is_surrounded_pattern(&grid, r, c, &['S', 'A', 'M'], 'S', 'M') {
                    p2 += 1;
                }
            }
        }
    }

    println!("{}", p1);
    println!("{}", p2);
}

fn is_pattern(
    grid: &Vec<String>,
    r: usize,
    c: usize,
    pattern: &[char],
    direction: (isize, isize),
) -> bool {
    for (i, &ch) in pattern.iter().enumerate() {
        let nr = r as isize + i as isize * direction.0;
        let nc = c as isize + i as isize * direction.1;
        if nr < 0 || nc < 0 || nr >= grid.len() as isize || nc >= grid[0].len() as isize {
            return false;
        }
        if grid[nr as usize].chars().nth(nc as usize).unwrap_or(' ') != ch {
            return false;
        }
    }
    true
}

fn is_surrounded_pattern(
    grid: &Vec<String>,
    r: usize,
    c: usize,
    pattern: &[char],
    top_left: char,
    bottom_right: char,
) -> bool {
    if !is_pattern(grid, r, c, pattern, (1, 1)) {
        return false;
    }
    let br = grid[r + 2].chars().nth(c).unwrap_or(' ');
    let tr = grid[r].chars().nth(c + 2).unwrap_or(' ');
    br == top_left && tr == bottom_right
}
